File: python/7-calculator-with-error-handling.py
# Simple Calculator with Error Handling
# This script demonstrates how to handle exceptions in Python

def add(x, y):
    """Add two numbers."""
    return x + y

def subtract(x, y):
    """Subtract y from x."""
    return x - y

def multiply(x, y):
    """Multiply two numbers."""
    return x * y

def divide(x, y):
    """
    Divide x by y.
    
    Raises:
        ZeroDivisionError: If y is zero
    """
    if y == 0:
        raise ZeroDivisionError("Division by zero is not allowed!")
    return x / y

def calculator():
    """Run the calculator application."""
    print("=== Simple Calculator ===")
    print("This calculator handles division by zero errors.")
    
    while True:
        print("\nOperations:")
        print("1. Add")
        print("2. Subtract")
        print("3. Multiply")
        print("4. Divide")
        print("5. Exit")
        
        choice = input("\nEnter your choice (1-5): ")
        
        if choice == '5':
            print("Goodbye!")
            break
            
        if choice not in ['1', '2', '3', '4']:
            print("Invalid choice. Please enter a number between 1 and 5.")
            continue
            
        try:
            num1 = float(input("Enter first number: "))
            num2 = float(input("Enter second number: "))
            
            if choice == '1':
                result = add(num1, num2)
                print(f"{num1} + {num2} = {result}")
                
            elif choice == '2':
                result = subtract(num1, num2)
                print(f"{num1} - {num2} = {result}")
                
            elif choice == '3':
                result = multiply(num1, num2)
                print(f"{num1} × {num2} = {result}")
                
            elif choice == '4':
                try:
                    result = divide(num1, num2)
                    print(f"{num1} ÷ {num2} = {result}")
                except ZeroDivisionError as e:
                    print(f"Error: {e}")
                    
        except ValueError:
            print("Error: Please enter valid numbers.")
        except Exception as e:
            print(f"An unexpected error occurred: {e}")

# Run the calculator
if __name__ == "__main__":
    calculator() 

File: javascript/7-calculator-with-error-handling.js
/**
 * Simple Calculator with Error Handling
 * 
 * A command-line calculator that demonstrates:
 * - Basic arithmetic operations
 * - Error handling with try-catch blocks
 * - Input validation
 * - Async/await for user input
 * - Switch statements
 * - Function organization
 * 
 * Features:
 * - Addition, subtraction, multiplication, and division
 * - Division by zero error handling
 * - Invalid input validation
 * - Continuous operation until user exits
 * - Clean error messages
 * 
 * @author Ibrahim
 * @version 1.0
 */

const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

/**
 * Adds two numbers
 * @param {number} x - First number
 * @param {number} y - Second number
 * @returns {number} The sum of x and y
 */
function add(x, y) {
    return x + y;
}

/**
 * Subtracts two numbers
 * @param {number} x - First number
 * @param {number} y - Second number
 * @returns {number} The difference between x and y
 */
function subtract(x, y) {
    return x - y;
}

/**
 * Multiplies two numbers
 * @param {number} x - First number
 * @param {number} y - Second number
 * @returns {number} The product of x and y
 */
function multiply(x, y) {
    return x * y;
}

/**
 * Divides two numbers
 * @param {number} x - First number
 * @param {number} y - Second number
 * @returns {number} The quotient of x divided by y
 * @throws {Error} When attempting to divide by zero
 */
function divide(x, y) {
    if (y === 0) {
        throw new Error("Division by zero is not allowed!");
    }
    return x / y;
}

/**
 * Prompts the user with a question and returns their response
 * @param {string} question - The question to ask the user
 * @returns {Promise<string>} The user's response
 */
async function askQuestion(question) {
    return new Promise((resolve) => {
        readline.question(question, resolve);
    });
}

/**
 * Main calculator function that handles the program flow
 */
async function calculator() {
    console.log("=== Simple Calculator ===");
    console.log("This calculator handles division by zero errors.");
    
    while (true) {
        console.log("\nOperations:");
        console.log("1. Add");
        console.log("2. Subtract");
        console.log("3. Multiply");
        console.log("4. Divide");
        console.log("5. Exit");
        
        const choice = await askQuestion("\nEnter your choice (1-5): ");
        
        if (choice === '5') {
            console.log("Goodbye!");
            readline.close();
            break;
        }
        
        if (!['1', '2', '3', '4'].includes(choice)) {
            console.log("Invalid choice. Please enter a number between 1 and 5.");
            continue;
        }
        
        try {
            const num1 = parseFloat(await askQuestion("Enter first number: "));
            const num2 = parseFloat(await askQuestion("Enter second number: "));
            
            if (isNaN(num1) || isNaN(num2)) {
                throw new Error("Please enter valid numbers.");
            }
            
            let result;
            switch (choice) {
                case '1':
                    result = add(num1, num2);
                    console.log(`${num1} + ${num2} = ${result}`);
                    break;
                    
                case '2':
                    result = subtract(num1, num2);
                    console.log(`${num1} - ${num2} = ${result}`);
                    break;
                    
                case '3':
                    result = multiply(num1, num2);
                    console.log(`${num1} × ${num2} = ${result}`);
                    break;
                    
                case '4':
                    try {
                        result = divide(num1, num2);
                        console.log(`${num1} ÷ ${num2} = ${result}`);
                    } catch (e) {
                        console.log(`Error: ${e.message}`);
                    }
                    break;
            }
        } catch (e) {
            console.log(`Error: ${e.message}`);
        }
    }
}

// Start the calculator
calculator().catch(console.error); 